query/transformation.go

```diff
@@ -0,0 +1,195 @@
+	result := api.SeriesList{
+		Series:    make([]api.Timeseries, len(list.Series)),
+		Timerange: list.Timerange,
+		Name:      list.Name,
+	}
+	var err error
+	for i, series := range list.Series {
+		result.Series[i], err = transformTimeseries(series, transform, parameters, list.Timerange)
+		if err != nil {
+			return api.SeriesList{}, err
+		}
+	}
+	return result, nil
+}
+
+// checkParameters is used to make sure that each transform is given the right number of parameters.
+func checkParameters(name string, expected int, parameters []value) error {
+	if len(parameters) != expected {
+		printArgs := append([]value{stringValue("(SeriesList)")}, parameters...)
+		return errors.New(fmt.Sprintf("expected %s to be given %d parameters but was given %d: %+v", name, expected+1, len(parameters)+1, printArgs))
+	}
+	return nil
+}
+
+// transformDerivative estimates the "change per second" between the two samples (scaled consecutive difference)
+func transformDerivative(values []float64, parameters []value, timerange api.Timerange) ([]float64, error) {
+	if err := checkParameters("transform.derivative", 0, parameters); err != nil {
+		return nil, err
+	}
+	result := make([]float64, len(values))
+	for i := range values {
+		if i == 0 {
+			// The first element has 0
+			result[i] = 0
+			continue
+		}
+		// Otherwise, it's the scaled difference
+		result[i] = (values[i] - values[i-1]) / timerangeScale(timerange)
+	}
+	return result, nil
+}
+
+// transformIntegral integrates a series whose values are "X per second" to estimate "total X so far"
+// if the series represents "X in this sampling interval" instead, then you should use transformCumulative.
+func transformIntegral(values []float64, parameters []value, timerange api.Timerange) ([]float64, error) {
+	if err := checkParameters("transform.integral", 0, parameters); err != nil {
+		return nil, err
+	}
+	result := make([]float64, len(values))
+	integral := 0.0
+	for i := range values {
+		integral += values[i]
+		result[i] = integral * timerangeScale(timerange)
+	}
+	return result, nil
+}
+
+// transformRate functions exactly like transformDerivative but bounds the result to be positive and does not normalize.
+// That is, it returns consecutive differences which are at least 0.
+func transformRate(values []float64, parameters []value, timerange api.Timerange) ([]float64, error) {
+	if err := checkParameters("transform.rate", 0, parameters); err != nil {
+		return nil, err
+	}
+	result := make([]float64, len(values))
+	for i := range values {
+		if i == 0 {
+			result[i] = 0
+			continue
+		}
+		result[i] = (values[i] - values[i-1]) / timerangeScale(timerange)
+		if result[i] < 0 {
+			result[i] = 0
+		}
+	}
+	return result, nil
+}
+
+// transformCumulative computes the cumulative sum of the given values.
+func transformCumulative(values []float64, parameters []value, timerange api.Timerange) ([]float64, error) {
```

Comment:
The resolution of a series can't be chosen as an argument, it's computed from the properties of the given serieslist. So, from a caller's perspective, there's no way to select it.

`transformIntegral` makes sense for things that automatically keep track of rates in units of time (genuine rates in units of events/second or similar) 

`transormCumulative` makes sense for metrics that have counts like "messages received since last sample" which don't take the actual passage of time into account.

(On a related note, I just realized that a sampling method would have to be "sum" rather than "average" to deal with that last case - otherwise the resolution you'd sample at would totally change how the data looks).